### DISCUSSION TITLE
Geostatistically reasonable parameter field with PESTPP-IES

### DISCUSSION THREAD

**Discussion by d-strom** [usgs/pestpp#293] [Q&A] [answered]
Hi,

I have been running different instances of pestpp for some time now. Looking at the results from some pestpp-ies runs I find it difficult to reach a parameter field that varies in a geostatistically reasonable way. The ies_reg_factor doesn't really make that much of a difference (which I suppose is expected, since it doesn't know anything about the geological materials in play).

I still find that there are sort of "bullet-point" behavior which I would like to avoid. What would be the work-flow to achieve smoother transitioning of for example k-values When it comes to pestpp-ies?

Obviously its a bit hard to tell without the scale for K. But i think the patterns tell enough about variability to get an idea of how they all preform.

Below are three examples of a k-parameter field from the Rocky Mounatains Arsenal example in Model Muse.

![image](https://github.com/usgs/pestpp/assets/74509984/61d8f777-5246-42cd-a811-beac7101543e)
Fig. 1. PESTPP-GLM, No regularization

![image](https://github.com/usgs/pestpp/assets/74509984/b870d119-fed8-4c15-a4fc-e9505e93eae9)
Fig. 2. PESTPP-GLM, first order pearson thikonov with geostatistical variogram

![image](https://github.com/usgs/pestpp/assets/74509984/8e843974-8875-40ed-8fce-e0f3a4c44e77)
Fig. 3. PESTPP-IES, ies_reg_factor = 0.25

![image](https://github.com/usgs/pestpp/assets/74509984/5801a1d4-d7fd-426b-b415-ed9284fc9f9b)
Fig. 4. Ordinary pest with thikonov regularization

**Comment by jtwhite79** [comment]
Did you bake the same geostat correlation structure into the prior parameter ensemble that is being used to enforce the regularization with pest/pestpp-glm?

**Reply by d-strom** [reply]
By that, do you mean something like running a prior monte carlo (ies noptmax -1) rum and setting up a geostat cov matrix and using as inputs for parcov(), ies_parameter_ensemble() and ies_observation_ensemble()?

**Reply by jtwhite79** [reply]
I mean did you use a geostatistical covariance matrix when you generated the prior parameter ensemble (one that has the same underlying geostatisical structure as used when enforcing a regularization preference for smoothness)? Have you visualized some of the prior parameter realizations - do they have the patterns and characteristics you want?

**Reply by d-strom** [reply]
I have not. I’m struggeling a bit with that since there are multiple sets of pilot points associated with multiple parameters (recharge, k, etc.) scattered over the domain. Not sure how to bake all of these into the workflow.
Since there seems to be only one cov matrix input?

**Reply by jtwhite79** [reply]
the one prior cov matrix can and should (and must!) include all adjustable parameters and can be block-diagonal with each block corresponding to the correlated parameters with in a given pilot point set.  Or you can make an pest-style unc file that also describes the same block-diagonal cov matrix.  Or if you have heaps of parameters, it can be more efficient and easier to generate the (correlated) prior parameter ensemble prior to runnning ies - this can be done with pyemu or the pest utils - in this case, you dont need to pass the prior cov matrix to ies...

**Reply by jtwhite79** [reply]
...anyhow, without baking the same preferred geostatistical relations into the prior parameter ensemble as used in tikhonov regularization with pest/pestpp-glm, I'm not surprised the ies result has those spatially isolated high and low values - its simply replicating/using the spatial patterns found in the prior parameter ensemble...

**Reply by d-strom** [reply]
I see.. I have generated prior covariance matrices with mkppstat and ppcov_sva. So the prior covariences are in place. Its just that I lack the tools to pass the individual covariance matrices to single block diagonal prior cov matrix. So yeah, I suppose I have the geostatistical patterns that are being used in the underlying pest run.

Does the pyemu suite or pypest utils offer some functionality to combine these prior covs into a single file (.unc-style perhaps?).

**Reply by cnicol-gwlogic** [reply, marked as answer]
You've done the hard yards already with mkppstat and ppcov - it'll take you about 5mins to build an unc file (just a text file that refers to those .mat files you made, and/or specified standard deviations for any unrelated params). See the original pest manual for the format, or [here](https://help.pesthomepage.org/parameter-uncertainty.html#parameteruncertaintyfile). Then just refer to that unc file in the ++parcov() entry in your control file.

**Reply by d-strom** [reply]
Cool. I’ll look into it tomorrow. Thanks for the advice! :)

**Reply by d-strom** [reply]
At first i thought the problem was solved by changing one of the parameters to logarithmic scale, but in the end it had nothing to do with it..

Reading through the link you provided among other help sections on pest homepage, it sure looks like the ppcov_sva could be used in par with the parcov() argument for pest++. Can they be passed through an .unc in their native format or do they need further preprocessing to match the requirements of pest++?

Since pest hangs with exception when the .unc-file is passed, something obviously goes wrong with the prior cov. Any ideas?

(Exception: run() returned non-zero: -1073741819)

**Reply by jtwhite79** [reply]
We need to see the actually stdout/stderr from the commandline/terminal and/or .rec file to see the error message

**Reply by d-strom** [reply]
The .rec file doesnt give many clues I'm afraid.
[rmamf6start1.txt](https://github.com/usgs/pestpp/files/15030303/rmamf6start1.txt)

![image](https://github.com/usgs/pestpp/assets/74509984/dd298637-8155-4a72-838d-8d3daacf9118)

It does, however, seem to recognize the covariance file. Since when I loaded in another file it told me the following:
![image](https://github.com/usgs/pestpp/assets/74509984/1253c1da-6c15-442b-9180-b42985d6423c)

Input from the python script below:
![image](https://github.com/usgs/pestpp/assets/74509984/6dd4835c-a161-4ba6-96bc-151b58e93768)


I'm running it from the gmdsi-tutorial environment if thats any help...

**Reply by cnicol-gwlogic** [reply]
Maybe post your unc file, there must be an issue in there. The format of the .mat files from ppcov are fine for ies.

**Reply by d-strom** [reply]
the unc-file as well as the cov-matrices are attatched in the post below from earlier. I have saved them as .txt to be able to upload them on here. cov.txt is my .unc-file. Should have mentioned before :)

**Reply by d-strom** [reply]
I have continued troubleshooting. This is what I have found out so far.
1. The parcov dosnt seem to accept the covariance matrices generated by ppcov_sva right off the bat if i just try to attatch one of them.
2. The If i generate a parcov with the code below, and add the prior.jcb to my parcov() statement. It runs with the expected variogram
    
    #read template file
    df_pp = pyemu.pp_utils.pp_tpl_to_dataframe(os.path.join(directory, filename))

    #make variogram
    v = pyemu.geostats.ExpVario(contribution=1.0, a=2000)
    gs = pyemu.geostats.GeoStruct(variograms=v, nugget=0.0)

    #create cov-matrix
    cov = gs.covariance_matrix(df_pp.x, df_pp.y, df_pp.parnme)
    cov.to_coo(os.path.join(tmp_d,"prior.jcb"))

3. If i add prior.jcb to my .unc file like below. And add the pass the .unc-file to the parcov() statement. it does not want to run...

    START COVARIANCE_MATRIX

     file "prior.jcb"
     variance_multiplier 1.0

    END COVARIANCE_MATRIX

So it seems like there might be some comparability problems with the matrices generated by ppcov_sva and .unc-files does not seem to be working with parcov (?), or the file-structure of the unc-file needs to be changed.

What are your thoughts?

**Reply by cnicol-gwlogic** [reply]
I use ppcov_sva mat files all the time via an unc file with ies. There is no compatibility issue. Something else is going on here with your setup. 

Btw I cannot see the unc file attachment you refer to.

**Reply by d-strom** [reply]
Alright. I'll attatch them again here (as .txt files). Well thats good news. Since it seems to be human error on my side of things, atleast I know its possible :). You find the files attatched to this message.

[RmaMf6Start.R_pp.R.1.Cov.txt](https://github.com/usgs/pestpp/files/15040163/RmaMf6Start.R_pp.R.1.Cov.txt) (cov 1)

[RmaMf6Start.Kx.K.1.Cov (1).txt](https://github.com/usgs/pestpp/files/15040159/RmaMf6Start.Kx.K.1.Cov.1.txt) (cov 2)

[cov.txt](https://github.com/usgs/pestpp/files/15040157/cov.txt) (unc file)

**Reply by d-strom** [reply]
uh oh... how silly. It was that the .unc file required 3 spaces instead of 2 before file etc.
I copied it from the website so i didnt think it could go wrong.
When copying from the pest manual it worked out in the end.

Thanks for your patience.

**Comment by d-strom** [comment]
> You've done the hard yards already with mkppstat and ppcov - it'll take you about 5mins to build an unc file (just a text file that refers to those .mat files you made, and/or specified standard deviations for any unrelated params). See the original pest manual for the format, or [here](https://help.pesthomepage.org/parameter-uncertainty.html#parameteruncertaintyfile). Then just refer to that unc file in the ++parcov() entry in your control file.

I took up on your advice and made the file cov.unc together with my covariance matrices (attatched). For whatever reason, when i load the .unc file into my parcov() control variable (pst.pestpp_options["parcov"] = os.path.join("cov.unc")). the master do not connect with the workers... It starts initializing the prior as the last step. It doesnt inform anything in particular in the rmr, or rec files regarding whats failing.

Can you see anything strange with the files that could cause this behavior? 

[cov.txt](https://github.com/usgs/pestpp/files/15022289/cov.txt)

[RmaMf6Start.R_pp.R.1.Cov.txt](https://github.com/usgs/pestpp/files/15022277/RmaMf6Start.R_pp.R.1.Cov.txt)
[RmaMf6Start.Kx.K.1.Cov.txt](https://github.com/usgs/pestpp/files/15022278/RmaMf6Start.Kx.K.1.Cov.txt)

![image](https://github.com/usgs/pestpp/assets/74509984/5f6dd2fb-ac3b-4c4e-b67c-79ef9236611e)

**Reply by cnicol-gwlogic** [reply]
You.may need to leave it work for a while to generate the ensemble. It will take longer with your prior cov than it did without any cov.

**Reply by d-strom** [reply]
I see, problem is that the master client crashes before it gets to work at all and removes all workers. I'm not sure how I could leave it to work then?

**Reply by d-strom** [reply]
It seems as if it doesnt read in the .unc-file correctly. I dont really see anything wrong with it and it should follow the conventions described in the pest manual. Can the fact that it contains non-numerical values play a role? the names of parameters that is.

I've tried to pass the .unc-file alone containing the covariances without any prior parameter- or observation ensembles.
